src/codecs/webp/decoder.rs

```diff
@@ -48,65 +47,80 @@ impl From<DecoderError> for ImageError {
 
 impl error::Error for DecoderError {}
 
-/// WebP Image format decoder. Currently only supports lossy RGB images.
-pub struct WebPDecoder<R> {
-    r: R,
-    frame: Frame,
-    have_frame: bool,
+enum InnerDecoder<R: Read> {
+    Lossy(Vp8Decoder<R>),
+    Lossless(LosslessDecoder<R>),
+}
+
+/// WebP Image format decoder. Currently only supports lossy RGB images or lossless RGBA images.
+pub struct WebPDecoder<R: Read> {
+    inner_decoder: InnerDecoder<R>,
 }
 
 impl<R: Read> WebPDecoder<R> {
     /// Create a new WebPDecoder from the Reader ```r```.
     /// This function takes ownership of the Reader.
-    pub fn new(r: R) -> ImageResult<WebPDecoder<R>> {
-        let f: Frame = Default::default();
-
-        let mut decoder = WebPDecoder {
-            r,
-            have_frame: false,
-            frame: f,
+    pub fn new(mut r: R) -> ImageResult<WebPDecoder<R>> {
+        
+        let _size = WebPDecoder::read_file_header(&mut r)?;
+
+        let inner_decoder = WebPDecoder::read_frame(r)?;
```

Comment:
This is probably the most contentious line, the rest of the PR seems fine. However, moving the point of error for a bad image, as well as the costs associated with reading it, into the constructor is quite a different. In particular we no longer have a fast path into reading merely the header and metadata.

Reply:
I did this at the start while hacking it together to work initially.
I did it this way because the original code had new() call decoder.read_metadata() which called self.read_frame() so it didn't seem to make much difference.
I could try moving it around.

Do we need this behaviour? :
```
let mut framedata = Vec::new();
self.r.by_ref().take(len as u64).read_to_end(&mut framedata)?;
let m = io::Cursor::new(framedata);
```

Reply:
I did not realize that the old version was following that pattern as well. The place is presumably fine then. The usage of `read_to_end` ensures that the full chunks has been consumed from the underlying reader in all cases, and `io::Cursor` just enables reading from that chunk.

It would be possible to restore that behavior by moving the read of `_len` from `LosslessDecoder::new` to `read_frame` and adding similar code there.

Reply:
The issue with that is then we have to change the inner decoder signature to ```InnerDecoder<Cursor<Vec<u8>>>``` which seems a little cumbersome.  
Not sure what the most idiomatic thing to do is, maybe instead of storing the decoder in the ```WebPDecoder```, we could copy their frames to the ```WebPDecoder``` and use them like the old decoder.rs does?

Reply:
Oh the frame copy is done so the vp8decoder is dropped after the frame is read; I guess that's behaviour we want to preserve?

Reply:
Ok, I made changes so the decoder acts like the old version, just with the enum for different frame types.

Reply:
Perfect, this makes it easier to keep track of all the changes introduced by the PR. Looks good to me then, if you want to change this then feel free to provide motivation and a separate PR :+1: